tools/wake-format/emitter.cpp

```diff
@@ -372,56 +372,27 @@ wcl::doc Emitter::walk_data(ctx_t ctx, CSTElement node) { return walk_placeholde
 
 wcl::doc Emitter::walk_def(ctx_t ctx, CSTElement node) {
   ASSERT_TOKEN(node, CST_DEF);
-  wcl::doc_builder bdr;
-  CSTElement child = node.firstChildElement();
-
-  // def
-  ASSERT_TOKEN(child, TOKEN_KW_DEF);
-  bdr.append("def");
-  child.nextSiblingElement();
-
-  // ws
-  ASSERT_TOKEN(child, TOKEN_WS);
-  bdr.append(space());
-  child.nextSiblingElement();
-
-  // CST_ID, CST_APPLY, TODO: others?
-  assert(!child.empty());
-  assert(child.id() == CST_ID || child.id() == CST_APP);
-  bdr.append(walk_node(ctx.sub(bdr), child));
-  child.nextSiblingElement();
-
-  // ws
-  ASSERT_TOKEN(child, TOKEN_WS);
-  bdr.append(space());
-  child.nextSiblingElement();
-
-  //  =
-  ASSERT_TOKEN(child, TOKEN_P_EQUALS);
-  bdr.append("=");
-  child.nextSiblingElement();
-
-  // optional ws/nl
-  CONSUME_WS_NL(child);
-
-  // rhs
-  {
-    wcl::doc doc = space().concat(walk_node(ctx.sub(bdr), child));
-    if (fits(bdr, ctx, doc) && !requires_nl(ctx, child)) {
-      bdr.append(doc);
-    } else {
-      ctx_t n_ctx = ctx.nest();
-      bdr.append(newline(n_ctx));
-      bdr.append(walk_node(n_ctx.sub(bdr), child));
-    }
-  }
-  child.nextSiblingElement();
-
-  // optional ws/nl
-  CONSUME_WS_NL(child);
 
-  assert(child.empty());
-  return std::move(bdr).build();
+  return formatter()
+      .token(TOKEN_KW_DEF)
+      .ws()
+      .walk({CST_ID, CST_APP}, [this](ctx_t ctx, CSTElement node) { return walk_node(ctx, node); })
+      .ws()
+      .token(TOKEN_P_EQUALS)
+      .consume_wsnl()
+      .escape([this](wcl::doc_builder& bdr, ctx_t ctx, CSTElement& node) {
+        wcl::doc doc = space().concat(walk_node(ctx.sub(bdr), node));
+        if (fits(bdr, ctx, doc) && !requires_nl(ctx, node)) {
```

Comment:
It would be a bit more efficent to do the requires_nl check first but I vote we leave this like this for now until we hit performance issues. If everything goes to plan with memoization we should be fine I think.

Reply:
Ack